pkg/distributors/contracts/RewardsScheduler.sol

```diff
@@ -0,0 +1,134 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+import "./interfaces/IMultiRewards.sol";
+
+// solhint-disable not-rely-on-time
+
+/**
+ * Scheduler for MultiRewards contract
+ */
+contract RewardsScheduler {
+    using SafeERC20 for IERC20;
+
+    IMultiRewards private immutable _multirewards;
+
+    constructor() {
+        _multirewards = IMultiRewards(msg.sender);
+    }
+
+    enum RewardStatus { UNINITIALIZED, PENDING, STARTED }
+
+    struct ScheduledReward {
+        IERC20 pool;
+        IERC20 rewardsToken;
+        uint256 startTime;
+        address rewarder;
+        uint256 amount;
+        RewardStatus status;
+    }
+
+    event RewardScheduled(
+        bytes32 rewardId,
+        address indexed scheduler,
+        IERC20 indexed pool,
+        IERC20 indexed rewardsToken,
+        uint256 startTime,
+        uint256 amount
+    );
+    event RewardStarted(
+        bytes32 rewardId,
+        address indexed scheduler,
+        IERC20 indexed pool,
+        IERC20 indexed rewardsToken,
+        uint256 startTime,
+        uint256 amount
+    );
+
+    mapping(bytes32 => ScheduledReward) private _rewards;
+
+    function getScheduledRewardInfo(bytes32 rewardId) external view returns (ScheduledReward memory reward) {
+        return _rewards[rewardId];
+    }
+
+    function startRewards(bytes32[] calldata rewardIds) external {
+        for (uint256 r; r < rewardIds.length; r++) {
+            bytes32 rewardId = rewardIds[r];
+            ScheduledReward memory scheduledReward = _rewards[rewardId];
+
+            require(scheduledReward.status == RewardStatus.PENDING, "Reward cannot be started");
+            require(scheduledReward.startTime <= block.timestamp, "Reward start time is in the future");
+
+            scheduledReward.rewardsToken.approve(address(_multirewards), scheduledReward.amount);
+            _multirewards.notifyRewardAmount(
+                scheduledReward.pool,
+                scheduledReward.rewardsToken,
+                scheduledReward.amount,
+                scheduledReward.rewarder
+            );
+            emit RewardStarted(
+                rewardId,
+                msg.sender,
+                scheduledReward.pool,
+                scheduledReward.rewardsToken,
+                scheduledReward.startTime,
+                scheduledReward.amount
+            );
+
+            _rewards[rewardId].status = RewardStatus.STARTED;
```

Comment:
Shouldn't we move this up a few lines, right after the `require status == pending` check, to avoid reentrancy issues?

Reply:
moving it, but, how would reentrancy happen?  This function only calls `multiRewards.notifyRewardAmount` and `multirewards` is an immutable contract which does not call this contract as part of `notifyRewardAmount`

Reply:
`notifyRewardAmounts` calls token functions, both directly and indirectly via the Vault, so there's a lot of potential for this.

Reply:
Ah, yes, I suppose a malicious token would be able to drain the balance with a reentrancy attack -I think that would only allow them to drain the token's own balance (still bad)

either way, the change is in